frontend/apps/reader/modules/readerhighlight.lua

```diff
@@ -2217,6 +2217,31 @@ function ReaderHighlight:onUpdateHoldPanRate()
     self:setupTouchZones()
 end
 
+function ReaderHighlight:updateHighlightPageNumbers()
+    local bookmarks = self.ui.bookmark.bookmarks
+    local highlights = {}
+    for i = #bookmarks, 1, -1 do
+        local bookmark = bookmarks[i]
+        if bookmark.highlighted then
+            local highlight = self:getHighlightByDatetime(bookmark.datetime)
+            if highlight then
+                local pageno = self.ui.paging and bookmark.page or self.document:getPageFromXPointer(bookmark.page)
+                if highlights[pageno] == nil then
+                    highlights[pageno] = {}
+                end
+                table.insert(highlights[pageno], highlight)
+            else -- orphan
+                table.remove(bookmarks, i)
```

Comment:
This feels a bit radical: if because of some technical mess/errors, you have a bookmark with highlights that is not referenced in "highlights" (so, I guess, just not drawn), you'll be removing the more valuable thingie (with possible user notes) in "bookmarks".
Let orphans be :)